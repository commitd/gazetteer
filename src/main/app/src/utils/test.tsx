import { StylesProvider } from '@material-ui/styles'
import {
  createHistory,
  createMemorySource,
  History,
  LocationProvider,
} from '@reach/router'
import { render, RenderOptions, RenderResult } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { GenerateId } from 'jss'
import React from 'react'
import { ThemeProvider } from '@committed/components'

const generateClassName: GenerateId = (rule, styleSheet) => {
  const prefix =
    styleSheet === undefined || styleSheet.options.classNamePrefix === undefined
      ? ''
      : styleSheet.options.classNamePrefix
  return `${prefix}-${rule.key}`
}

const LightTheme: React.FC<{ readonly children?: React.ReactNode }> = ({
  children,
}) => (
  <StylesProvider generateClassName={generateClassName}>
    <ThemeProvider choice="light">{children}</ThemeProvider>
  </StylesProvider>
)

const DarkTheme: React.FC<{ readonly children?: React.ReactNode }> = ({
  children,
}) => (
  <StylesProvider generateClassName={generateClassName}>
    <ThemeProvider choice="dark">{children}</ThemeProvider>
  </StylesProvider>
)

export const renderPlain = render

export const renderLight = (
  ui: Readonly<React.ReactElement>,
  options?: Omit<RenderOptions, 'queries'>
): RenderResult => render(ui, { wrapper: LightTheme, ...options })

export const renderDark = (
  ui: Readonly<React.ReactElement>,
  options?: Omit<RenderOptions, 'queries'>
): RenderResult => render(ui, { wrapper: DarkTheme, ...options })

export function renderWithRouter(
  ui: Readonly<React.ReactElement>,
  route = '/',
  options?: Omit<RenderOptions, 'queries'>,
  delegate = renderLight
): RenderResult & { history: History } {
  const history = createHistory(createMemorySource(route))
  return {
    ...delegate(
      <LocationProvider history={history}>{ui}</LocationProvider>,
      options
    ),
    history,
  }
}
// re-export everything
export * from '@testing-library/react'
// override render method
export { renderLight as render }
export { userEvent }
