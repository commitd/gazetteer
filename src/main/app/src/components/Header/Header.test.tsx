import { Icons } from '@committed/components'
import React from 'react'
import { Header, MainAction, ToolbarAction, ToolbarSelect } from '.'
import { render, renderDark, userEvent } from '../../utils/test'

// NB
// We don't snapshot some of these as we found that sometimes the snapshots have a touch ripple element and sometimes they don't.
// I think this is a timing issue, but it meant some tests failed occasionally.
// Instead we just check there is some output rather than have sporadically failing tests.

it('Header renders without error', () => {
  const { asFragment } = render(<Header />)
  expect(asFragment()).toMatchSnapshot()
})

it('Header renders dark without error', () => {
  const { asFragment } = renderDark(<Header />)
  expect(asFragment()).toMatchSnapshot()
})

it('MainAction renders without error', () => {
  const onClick = jest.fn()
  const { asFragment } = render(
    <MainAction onClick={onClick}>Action</MainAction>
  )
  expect(asFragment()).toMatchSnapshot()
})

it('MainAction renders dark without error', () => {
  const onClick = jest.fn()
  const { asFragment } = renderDark(
    <MainAction onClick={onClick}>Action</MainAction>
  )
  expect(asFragment()).toBeDefined()
})

it('MainAction renders with icon without error', () => {
  const onClick = jest.fn()
  const { asFragment } = render(
    <MainAction onClick={onClick} icon={<Icons.Star />}>
      Action
    </MainAction>
  )
  expect(asFragment()).toBeDefined()
})

it('MainAction onClick is called', () => {
  const onClick = jest.fn()
  const { getByRole } = render(
    <MainAction onClick={onClick} icon={<Icons.Star />}>
      Action
    </MainAction>
  )
  expect(onClick).not.toBeCalled()
  userEvent.click(getByRole('button', { name: /Action/i }))
  expect(onClick).toBeDefined()
})

it('ToolbarAction renders without error', () => {
  const onClick = jest.fn()
  const { asFragment } = render(
    <ToolbarAction onClick={onClick}>Action</ToolbarAction>
  )
  expect(asFragment()).toBeDefined()
})

it('ToolbarAction renders dark without error', () => {
  const onClick = jest.fn()
  const { asFragment } = renderDark(
    <ToolbarAction onClick={onClick}>Action</ToolbarAction>
  )
  expect(asFragment()).toBeDefined()
})

it('ToolbarAction renders with icon', () => {
  const onClick = jest.fn()
  const { asFragment } = renderDark(
    <ToolbarAction onClick={onClick} icon={<Icons.Save />}>
      Action
    </ToolbarAction>
  )
  expect(asFragment()).toBeDefined()
})

it('ToolbarAction click calls callback', () => {
  const onClick = jest.fn()
  const { getByRole } = renderDark(
    <ToolbarAction onClick={onClick} icon={<Icons.Save />}>
      Action
    </ToolbarAction>
  )

  expect(onClick).not.toBeCalled()
  userEvent.click(getByRole('button', { name: /Action/i }))
  expect(onClick).toHaveBeenCalled()
})

it('ToolbarSelect renders without error', () => {
  const onClick = jest.fn()
  const { asFragment } = render(
    <ToolbarSelect
      selectedIndex={0}
      onClick={onClick}
      options={['Option 1', 'Option 2']}
    />
  )
  expect(asFragment()).toBeDefined()
})

it('ToolbarSelect renders dark without error', () => {
  const onClick = jest.fn()
  const { asFragment } = renderDark(
    <ToolbarSelect
      selectedIndex={0}
      onClick={onClick}
      options={['Option 1', 'Option 2']}
    />
  )
  expect(asFragment()).toBeDefined()
})

it('ToolbarSelect renders placeholder', () => {
  const onClick = jest.fn()
  const { getByText } = render(
    <ToolbarSelect
      selectedIndex={-1}
      onClick={onClick}
      placeholder="Test"
      options={['Option 1', 'Option 2']}
    />
  )
  getByText('Test')
})

it('ToolbarSelect shows selected option', () => {
  const onClick = jest.fn()
  const { getByText } = render(
    <ToolbarSelect
      selectedIndex={1}
      onClick={onClick}
      placeholder="Test"
      options={['Option 1', 'Option 2']}
    />
  )
  getByText('Option 2')
})

it('ToolbarSelect can select another option', () => {
  const onClick = jest.fn()
  const options = ['Option 1', 'Option 2']
  const { getByRole, getByText } = render(
    <ToolbarSelect selectedIndex={0} onClick={onClick} options={options} />
  )

  expect(onClick).not.toBeCalled()
  userEvent.click(getByRole('button', { name: options[0] }))
  expect(onClick).not.toBeCalled()
  userEvent.click(getByText(options[1]))
  expect(onClick).toHaveBeenCalledWith(1)
})
