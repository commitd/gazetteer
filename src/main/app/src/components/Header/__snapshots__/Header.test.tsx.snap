// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders dark without error 1`] = `
<DocumentFragment>
  <div>
    <header
      class="MuiPaper-root MuiAppBar-root MuiAppBar-positionSticky MuiAppBar-colorPrimary WithStyles(ForwardRef(AppBar))-root WithStyles(ForwardRef(AppBar))-root Styled(WithStyles(ForwardRef(AppBar)))-root Styled(WithStyles(ForwardRef(AppBar)))-root MuiPaper-elevation0"
    >
      <div
        class="MuiContainer-root MuiContainer-maxWidthLg"
      >
        <div
          class="MuiToolbar-root MuiToolbar-regular Component-root Component-root MuiToolbar-gutters"
        >
          <div
            class="MuiBox-root MuiBox-root Styled(MuiBox)-root Styled(MuiBox)-root"
          >
            <a
              aria-current="page"
              class="Link-root"
              href="/"
            >
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-text WithNoOpener-root WithNoOpener-root"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 8c-2.21 0-4 1.79-4 4s1.79 4 4 4 4-1.79 4-4-1.79-4-4-4zm8.94 3c-.46-4.17-3.77-7.48-7.94-7.94V1h-2v2.06C6.83 3.52 3.52 6.83 3.06 11H1v2h2.06c.46 4.17 3.77 7.48 7.94 7.94V23h2v-2.06c4.17-.46 7.48-3.77 7.94-7.94H23v-2h-2.06zM12 19c-3.87 0-7-3.13-7-7s3.13-7 7-7 7 3.13 7 7-3.13 7-7 7z"
                    />
                  </svg>
                  <h1
                    class="MuiTypography-root WithStyles(ForwardRef(Typography))-root WithStyles(ForwardRef(Typography))-root Component-root Component-root MuiTypography-h1"
                  >
                     Gazetteer
                  </h1>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </a>
          </div>
          <button
            aria-label="switch-theme"
            class="MuiButtonBase-root MuiIconButton-root WithNoOpener-root WithNoOpener-root MuiIconButton-colorPrimary"
            tabindex="0"
            title="Use dark theme"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root makeStyles-color makeStyles-color"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M9 2c-1.05 0-2.05.16-3 .46 4.06 1.27 7 5.06 7 9.54s-2.94 8.27-7 9.54c.95.3 1.95.46 3 .46 5.52 0 10-4.48 10-10S14.52 2 9 2z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </header>
  </div>
</DocumentFragment>
`;

exports[`Header renders without error 1`] = `
<DocumentFragment>
  <div>
    <header
      class="MuiPaper-root MuiAppBar-root MuiAppBar-positionSticky MuiAppBar-colorPrimary WithStyles(ForwardRef(AppBar))-root WithStyles(ForwardRef(AppBar))-root Styled(WithStyles(ForwardRef(AppBar)))-root Styled(WithStyles(ForwardRef(AppBar)))-root MuiPaper-elevation0"
    >
      <div
        class="MuiContainer-root MuiContainer-maxWidthLg"
      >
        <div
          class="MuiToolbar-root MuiToolbar-regular Component-root Component-root MuiToolbar-gutters"
        >
          <div
            class="MuiBox-root MuiBox-root Styled(MuiBox)-root Styled(MuiBox)-root"
          >
            <a
              aria-current="page"
              class="Link-root"
              href="/"
            >
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-text WithNoOpener-root WithNoOpener-root"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 8c-2.21 0-4 1.79-4 4s1.79 4 4 4 4-1.79 4-4-1.79-4-4-4zm8.94 3c-.46-4.17-3.77-7.48-7.94-7.94V1h-2v2.06C6.83 3.52 3.52 6.83 3.06 11H1v2h2.06c.46 4.17 3.77 7.48 7.94 7.94V23h2v-2.06c4.17-.46 7.48-3.77 7.94-7.94H23v-2h-2.06zM12 19c-3.87 0-7-3.13-7-7s3.13-7 7-7 7 3.13 7 7-3.13 7-7 7z"
                    />
                  </svg>
                  <h1
                    class="MuiTypography-root WithStyles(ForwardRef(Typography))-root WithStyles(ForwardRef(Typography))-root Component-root Component-root MuiTypography-h1"
                  >
                     Gazetteer
                  </h1>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </a>
          </div>
          <button
            aria-label="switch-theme"
            class="MuiButtonBase-root MuiIconButton-root WithNoOpener-root WithNoOpener-root MuiIconButton-colorPrimary"
            tabindex="0"
            title="Use dark theme"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root makeStyles-color makeStyles-color"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M9 2c-1.05 0-2.05.16-3 .46 4.06 1.27 7 5.06 7 9.54s-2.94 8.27-7 9.54c.95.3 1.95.46 3 .46 5.52 0 10-4.48 10-10S14.52 2 9 2z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </header>
  </div>
</DocumentFragment>
`;

exports[`MainAction renders without error 1`] = `
<DocumentFragment>
  <div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained WithNoOpener-root WithNoOpener-root MuiButton-containedPrimary"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Action
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</DocumentFragment>
`;
